dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/icon/DefaultIconService.java

```diff
@@ -57,141 +53,139 @@
-    return defaultIcons.get(key) != null || customIconStore.getIconByKey(key) != null;
+    return iconStore.getIconByKey(key) != null;
   }
 
   @Override
   @Transactional
-  public void addCustomIcon(CustomIcon customIcon) throws BadRequestException, NotFoundException {
-    validateIconDoesNotExists(customIcon.getKey());
-    FileResource fileResource = getFileResource(customIcon.getFileResourceUid());
-    UserDetails currentUserDetails = CurrentUserUtil.getCurrentUserDetails();
-    customIconStore.save(customIcon, fileResource, currentUserDetails);
-  }
+  public void addIcon(Icon icon) throws BadRequestException, NotFoundException, SQLException {
 
-  @Override
-  @Transactional
-  public void updateCustomIcon(String key, String description, String[] keywords)
-      throws BadRequestException, NotFoundException {
-    CustomIcon icon = validateCustomIconExists(key);
+    if (icon == null) {
+      throw new BadRequestException("Icon cannot be null.");
+    }
 
-    if (description == null && keywords == null) {
-      throw new BadRequestException(
-          String.format(
-              "Can't update icon %s if none of description and keywords are present in the request",
-              key));
+    validateIconDoesNotExists(icon);
+    validateIconKey(icon.getKey());
+
+    if (Boolean.TRUE.equals(icon.getCustom())) {
+      FileResource fileResource = getFileResource(icon.getFileResource().getUid());
+      fileResource.setAssigned(true);
+      fileResourceService.updateFileResource(fileResource);
     }
 
-    if (description != null) {
-      icon.setDescription(description);
+    if (icon.getCreatedBy() == null) {
+      icon.setCreatedBy(userService.getUserByUsername(CurrentUserUtil.getCurrentUsername()));
     }
 
-    if (keywords != null) {
-      icon.setKeywords(keywords);
+    if (icon.getKeywords() == null) {
+      icon.setKeywords(Set.of());
     }
 
     icon.setAutoFields();
-    customIconStore.update(icon);
+    iconStore.save(icon);
   }
 
   @Override
   @Transactional
-  public void updateCustomIcon(CustomIcon customIcon)
-      throws BadRequestException, NotFoundException {
-    updateCustomIcon(customIcon.getKey(), customIcon.getDescription(), customIcon.getKeywords());
+  public void updateIcon(Icon icon) throws BadRequestException, SQLException {
+
+    if (icon == null) {
+      throw new BadRequestException("Icon cannot be null.");
+    }
+
+    validateIconKeyNotNullOrEmpty(icon.getKey());
+
+    if (icon.getKeywords() == null) {
+      icon.setKeywords(Set.of());
+    }
+
+    icon.setAutoFields();
+    iconStore.update(icon);
   }
 
   @Override
   @Transactional
-  public void deleteCustomIcon(String key) throws BadRequestException, NotFoundException {
-    CustomIcon icon = validateCustomIconExists(key);
-    getFileResource(icon.getFileResourceUid()).setAssigned(false);
-    customIconStore.delete(key);
+  public void deleteIcon(Icon icon) throws BadRequestException, NotFoundException {
```

Comment:
Would `deleteIcon(String key)` simplify things? I imagine that is all we need to identify the icon.